bitcoin/src/taproot.rs

```diff
@@ -577,23 +792,129 @@ impl NodeInfo {
     }
 }
 
+impl TryFrom<TaprootBuilder> for NodeInfo {
+    type Error = IncompleteBuilder;
+
+    fn try_from(builder: TaprootBuilder) -> Result<Self, Self::Error> {
+        builder.try_into_node_info()
+    }
+}
+
+#[cfg(feature = "serde")]
+#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
+impl serde::Serialize for NodeInfo {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        use serde::ser::SerializeSeq;
+        let mut seq = serializer.serialize_seq(Some(self.leaves.len()))?;
+        for tap_leaf in self.leaves.iter() {
+            seq.serialize_element(&tap_leaf.merkle_branch().len())?;
+            seq.serialize_element(&tap_leaf.leaf)?;
+        }
+        seq.end()
+    }
+}
+
+#[cfg(feature = "serde")]
+#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
+impl<'de> serde::Deserialize<'de> for NodeInfo {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct SeqVisitor;
+        impl<'de> serde::de::Visitor<'de> for SeqVisitor {
+            type Value = NodeInfo;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("Taproot tree in DFS walk order")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: serde::de::SeqAccess<'de>,
+            {
+                let mut builder = TaprootBuilder::new();
```

Comment:
Would it make sense to have `fn with_capacity(capacity: usize) -> TaprootBuilder` and call it here? I suppose we would need to internally allocate `Vec` with `log2(capacity + 1)` (`size_of::<usize>() * 8 - (capacity + 1).leading_zeros()` if I'm not mistaken)

Reply:
I don't think we have to optimize for unoptimal trees - they will have bigger problems than a vec reallocating too many times.

Reply:
Ah right. capacity is just a good enough guess, does not have to be strict upper bound.

Reply:
Done. I don't think there is a +1 here. Used `size_of::<usize>() * 8 - (capacity).leading_zeros()` instead of `capacity + 1`